File: src/python/_0443_String_Compression.py
class Solution:
    def compress(self, chars: List[str]) -> int:
        turtle, idx = 0, 0

        for hare in range(1, len(chars) + 1):
            if hare < len(chars) and chars[turtle] == chars[hare]:
                continue

            count = hare - turtle
            chars[idx] = chars[turtle]
            idx += 1
            turtle = hare

            if count == 1:
                continue

            for digit in str(count):
                chars[idx] = digit
                idx += 1

        return idx

File: src/_0443_String_Compression.java
class Solution {
    public int compress(char[] chars) {
        int turtle = 0, idx = 0;

        for (int hare = 1; hare <= chars.length; hare++) {
            if (hare < chars.length && chars[turtle] == chars[hare])
                continue;

            int count = hare - turtle;
            chars[idx++] = chars[turtle];
            turtle = hare;

            if (count == 1)
                continue;

            if (count < 10) {
                chars[idx++] = (char) (count + '0');
            }

            for (char digit : Integer.toString(count).toCharArray())
                chars[idx++] = digit;

        }

        return idx;
    }
}
